doc/source/rllib/connector-v2.rst

```diff
@@ -0,0 +1,596 @@
+back into the episode, the subsequent module-to-env pipeline is able to see this change.
+The same is true for the Learner pipeline: The episodes that it operates on
+have already been through both env-to-module and module-to-env pipelines and might have been changed by these.
+
+
+Writing more complex custom ConnectorV2s
+----------------------------------------
+
+Besides simply transforming one observation at a time in a preprocessor-style setup as shown preceding,
+here are four more examples showcasing more complex customizations of the different :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` pipelines:
+
+* How to add rewards and or actions to the batch.
+* How to stack the N most recent observations.
+* How to manipulate the action logits coming from the :py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`.
+
+
+Adding past rewards to the Model's input
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you want to add the most recent reward to be part of the next observation going into your
+:py:class:`~ray.rllib.core.rl_module.rl_module.RLModule`, you can write a custom
+env-to-module :py:class:`~ray.rllib.connectors.connector_v2.ConnectorV2` piece like so:
+
+
+.. testcode::
+
+    import gymnasium as gym
+    import numpy as np
+    from ray.rllib.connectors.connector_v2 import ConnectorV2
+
+
+    class AddLastRewardToObservations(ConnectorV2):
+
+        # Define how the observation space will change because of this connector piece.
+        def recompute_output_observation_space(self, in_obs_space, in_act_space):
+            # For simplicity, assert input obs space is a 1D Box
+            assert isinstance(in_obs_space, gym.spaces.Box) and len(in_obs_space.shape) == 1
+            return gym.spaces.Box(
+                low=list(in_obs_space.low) + [float("-inf")],
+                high=list(in_obs_space.high) + [float("inf")],
+                shape=(in_obs_space.shape[0] + 1,),
+                dtype=in_obs_space.dtype,
+            )
+
+        # Define the actual transformation.
+        def __call__(self, *, rl_module, batch, episodes, **kwargs):
+            # Loop through all `episodes`.
+            for single_agent_episode in self.single_agent_episode_iterator(episodes):
+                # Get last reward (or 0.0 if right after `env.reset`).
+                reward = single_agent_episode.get_reward(-1, fill=0.0)
+                # Get last observation.
+                obs = single_agent_episode.get_observation(-1)
+                # Append reward to obs array.
+                new_obs = np.append(obs, reward)
+
+                # Write new observation back into the episode.
+                single_agent_episode.set_observation(new_value=new_obs, at_index=-1)
+
+            return batch
+
+
+.. tip::
+    There is already an off-the-shelf ConnectorV2 piece available for users, which performs the task of
```

Comment:
Just checking: are you saying that the above example implements this from scratch instead of using the built-in? Why not just use the built-in and reduce the code complexity for your example?